Moves alternate bases schema mutation to ProcessedVariantFactory.

bashir2 opened the pull request:
Tested: Ran integration tests and updated unit-tests. Also ran the code manually for an annotated VCF and confirmed the generated schema is the same as before.

Issue: #59, #81.

bashir2 commented:
Folks, please note that this needs some more unit-tests, which I will add tomorrow. No new functionality is added so the unit-test coverage is not decreased but some functionality that before were buried as module "private"s are now exposed as "public" interfaces so it is better to write tests for them in isolation (e.g., all of bigquery_util and the new schema method of ProcessedVariantFactory).

Also we probably need a better representation of the schema as an encapsulated class but that is left for the schema validation work that Nima is going to do as he can better design that class knowing the current status and what is needed for validation. So this does not "fix" Issue #59, just make it better.

arostamianfar commented on gcp_variant_transforms/libs/processed_variant.py:
```diff
@@ -261,6 +266,67 @@ def _create_annotation_map(self, annotations, annotation_names):
+    # type: (str) -> bool
+    if info_field_name not in self._header_fields.infos:
+      raise ValueError('INFO field {} not found'.format(info_field_name))
+    if ((self._split_alternate_allele_info_fields and
```
nit: you can just do return (...)

bashir2 replied:
Done.

bashir2 replied:
FYI, I changed this a little because pylint was suggesting to do it like [this](https://travis-ci.org/bashir2/gcp-variant-transforms/builds/344996460) (check the bottom of the page) and it did not seem to be very readable. So instead I do it the way it is now.

arostamianfar commented on gcp_variant_transforms/libs/bigquery_vcf_schema.py:
```diff
@@ -84,130 +41,89 @@ class _TableFieldConstants(object):
-          description=''))
-    alternate_bases_record.fields.append(annotation_record)
-  schema.fields.append(alternate_bases_record)
+  proc_var_factory.add_alt_record_to_schema(schema)
```
nit: instead of adding alt to a mutable schema, consider returning the alt schema and appending it to the schema object here (e.g. `schema.fields.append(proc_var_factory.get_alt_record_schema())`. This makes is more clear how the schema is being mutated in this function.

bashir2 replied:
Done.

arostamianfar commented on gcp_variant_transforms/transforms/variant_to_bigquery.py:
```diff
@@ -41,39 +43,38 @@ def process(self, record):
+        The latter functionality is what is needed here.
+      append: If true, existing records in output_table will not be
         overwritten. New records will be appended to those that already exist.
       omit_empty_sample_calls (bool): If true, samples that don't have a given
```
nit: can you also remove bool from here? :)

bashir2 replied:
Done.

arostamianfar commented on gcp_variant_transforms/transforms/variant_to_bigquery.py:
```diff
@@ -41,39 +43,38 @@ def process(self, record):
 class VariantToBigQuery(beam.PTransform):
   """Writes PCollection of ``Variant`` records to BigQuery."""
```
Please also change this to be ProcessedVariant.

bashir2 replied:
Done.

bashir2 commented:
BTW, if you want to see the changes since last review, check [this commit](https://github.com/bashir2/gcp-variant-transforms/commit/e64752c2777524a61ee173314845fea78dbef0fb)

bashir2 commented:
bigquery_util_test added plus a TODO for the new create_alt_record_for_schema method of ProcessedVariantFactory.

bashir2 commented on gcp_variant_transforms/libs/processed_variant.py:
```diff
@@ -261,6 +266,67 @@ def _create_annotation_map(self, annotations, annotation_names):
       annotation_dict[name] = annotations[index + 1]
     return annotation_dict
 
+  def create_alt_record_for_schema(self):
```
Changed it to `create_alt_bases_field_schema` because I want to make it clear it is not a full schema.

bashir2 commented on gcp_variant_transforms/libs/bigquery_vcf_schema.py:
```diff
@@ -84,130 +41,89 @@ class _TableFieldConstants(object):
-                                       split_alternate_allele_info_fields=True,
-                                       annotation_fields=None):
+def generate_schema_from_header_fields(
+    header_fields,  # type: vcf_header_parser.HeaderFields
```
Indeed :-)

bashir2 commented on gcp_variant_transforms/libs/bigquery_vcf_schema.py:
```diff
@@ -84,130 +41,89 @@ class _TableFieldConstants(object):
-                                       annotation_fields=None):
+def generate_schema_from_header_fields(
+    header_fields,  # type: vcf_header_parser.HeaderFields
+    proc_var_factory,  # type: processed_variant.ProcessedVariantFactory
```
Changed it to `proc_variant_factory`; agreed that `proc` is not great but `processed_variant_factory` is also too long (for example I need to break the `# type` comment to next line which is doable but ugly). There is a docstring below which clarifies it too. Still if you feel strongly about this, please let me know.

bashir2 commented on gcp_variant_transforms/libs/bigquery_vcf_schema.py:
```diff
@@ -84,130 +41,89 @@ class _TableFieldConstants(object):
+      VCF file(s).
+    proc_var_factory: The factory class that knows how to convert Variant
+      instances to ProcessedVariant. As a side effect it also knows how to
+      modify BigQuery schema based on the ProcessedVariants that it generates.
```
Agreed this is not great but I can't think of a much better alternative design without too much code change either. If you have a specific suggestion, I can add a TODO.

bashir2 commented:
BTW, if you want to see only the changes since last time, check [this commit](https://github.com/bashir2/gcp-variant-transforms/commit/64697b0082d428bb388d5bc59c11b061abef59dc).
